lib/Dialect/DC/DCOps.cpp

```diff
@@ -0,0 +1,240 @@
+  FuncOp fn = symbolTable.lookupNearestSymbolFrom<FuncOp>(*this, fnAttr);
+  if (!fn)
+    return emitOpError() << "'" << fnAttr.getValue()
+                         << "' does not reference a valid function";
+
+  // Verify that the operand and result types match the callee.
+  auto fnType = fn.getFunctionType();
+  if (fnType.getNumInputs() != getNumOperands())
+    return emitOpError("incorrect number of operands for callee");
+
+  for (unsigned i = 0, e = fnType.getNumInputs(); i != e; ++i)
+    if (getOperand(i).getType() != fnType.getInput(i))
+      return emitOpError("operand type mismatch: expected operand type ")
+             << fnType.getInput(i) << ", but provided "
+             << getOperand(i).getType() << " for operand number " << i;
+
+  if (fnType.getNumResults() != getNumResults())
+    return emitOpError("incorrect number of results for callee");
+
+  for (unsigned i = 0, e = fnType.getNumResults(); i != e; ++i)
+    if (getResult(i).getType() != fnType.getResult(i)) {
+      auto diag = emitOpError("result type mismatch at index ") << i;
+      diag.attachNote() << "      op result types: " << getResultTypes();
+      diag.attachNote() << "function result types: " << fnType.getResults();
+      return diag;
+    }
+
+  return success();
+}
+
+FunctionType CallOp::getCalleeType() {
+  return FunctionType::get(getContext(), getOperandTypes(), getResultTypes());
+}
+
+// =============================================================================
+// JoinOp
+// =============================================================================
+
+void JoinOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                         MLIRContext *context) {}
+
+// =============================================================================
+// ForkOp
+// =============================================================================
+
+template <typename TInt>
+static ParseResult parseIntInSquareBrackets(OpAsmParser &parser, TInt &v) {
+  if (parser.parseLSquare() || parser.parseInteger(v) || parser.parseRSquare())
+    return failure();
+  return success();
+}
+
+ParseResult ForkOp::parse(OpAsmParser &parser, OperationState &result) {
+  SmallVector<OpAsmParser::UnresolvedOperand, 4> operands;
+  llvm::SMLoc allOperandLoc = parser.getCurrentLocation();
+  size_t size = 0;
+  if (parseIntInSquareBrackets(parser, size))
+    return failure();
+
+  if (size == 0)
+    return parser.emitError(parser.getNameLoc(),
+                            "fork size must be greater than 0");
+
+  if (parser.parseOperandList(operands) ||
+      parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  auto tt = dc::TokenType::get(parser.getContext());
+  llvm::SmallVector<Type> operandTypes{tt};
+  SmallVector<Type, 1> resultTypes{size, tt};
+  result.addTypes(resultTypes);
+  if (parser.resolveOperands(operands, operandTypes, allOperandLoc,
```

Comment:
Likewise, use `parser.resolveOperand` (non-plural) if my understanding about fork is correct.